atlasdb-config/src/main/java/com/palantir/atlasdb/factory/LocalOrRemoteProxy.java

```diff
@@ -58,6 +59,10 @@ private LocalOrRemoteProxy(T localService, T remoteService, Future<Boolean> useL
     @Override
     protected Object handleInvocation(Object proxy, Method method, Object[] args) throws Throwable {
         while (true) {
+            if (Thread.currentThread().isInterrupted()) {
+                // do not clear the interrupt flag
+                throw new SafeIllegalStateException("interrupted");
```

Comment:
May be a misunderstanding: In the event that a `LocalOrRemoteProxy` is interrupted between iterations
(e.g. during the sleep), we will throw a `SafeIllegalStateException` instead of `InterruptedException`, and the interrupt flag is still set so the next interruptible thing will throw an `InterruptedException`. This feels a bit strange - is this what we want?

I'm thinking something like

```
            if (Thread.currentThread().isInterrupted()) {
                // do not clear the interrupt flag
                if (Arrays.stream(method.getExceptionTypes()).anyMatch(p -> p == InterruptedException.class)) {
                    throw new InterruptedException("interrupted");
                }
                throw new SafeIllegalStateException("interrupted");
```

Reply:
Yes, that is how thread interruption is supposed to work. You want the next blocking call to throw an InterruptedException. Something interrupted this thread and that interrupt has not been handled yet.

Whether we throw an InterruptedException or another type of exception + a set interrupt flag is not super important. Both ultimately result in an InterruptedException and signal to our caller that an interrupt has occurred.

Reply:
Yep. I didn't think it was super important (the basis of the above comment is basically whether we want `InterruptedException` or another type + set flag)

Reply:
I'd probably have written the alternative method, though not a big point as mentioned